### ISSUE TITLE
[java] UnusedPrivateMethod detected when class is referenced in another class

### ISSUE THREAD

**Issue by manuel-a-romeiro-alb** [pmd/pmd#5486]
**Affects PMD Version: 7.9.0**

**Rule: UnusedPrivateMethod**

**Rule documentation:**
https://docs.pmd-code.org/snapshot/pmd_rules_java_bestpractices.html#unusedprivatemethod

**Description:**
False positive when a private method is invoked directly over the return value of a method on other class.
The workaround is declaring a variable with the return value, and call the private method over that variable.

**Code Sample demonstrating the issue:**

```java
package foo;

public class Class1 {

    public void publicMethod() {
        new Class2().getClass1().privateMethod();
    }

    private void privateMethod() { // This method is detected as UnusedPrivateMethod (false positive)
        // do stuff
    }

    public void publicMethod2() {
        // Declaring the variable, makes the PMD works fine (workaround)
        Class1 class1 = new Class2().getClass1();
        class1.privateMethod2();
    }

    private void privateMethod2() {
        // do stuff
    }
}
```
```java
package foo;

public class Class2 {
    public Class1 getClass1() {
        return new Class1();
    }
}
```

**Expected outcome:**
No errors

**PMD log:**
[INFO] PMD version: 7.9.0                                                                                                    
[INFO] PMD Failure: foo.Class1:9 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'privateMethod()'..

**Running PMD through: Maven**

**Comment by oowekyala** [comment]
Hi, this is not related to the structure of the code. There is a problem with how you call PMD. Apparently the compiled class Class2 is not compiled or not the classpath.

When you declare a variable with an explicit type we know that it is Class1::privateMethod2 that is called. When you don't since we don't know the return type of Class2::getClass1 (because of missing class file), we don't know whether the privateMethod method is from Class1 or from another class.

---

Several recent issues with UnusedPrivateMethod and other rules fall in this category. There might be something wrong with the Maven pmd plugin @adangel (just FYI). Maybe if we detect missing types we should print a warning to the log (I think, we can do that in core though, we don't have to do it in pmd-maven-plugin).

**Comment by adangel** [comment]
Refs #5064 - in case the underlying problem is indeed missing types, we have already an issue to report this somehow.

**Comment by manuel-a-romeiro-alb** [comment]
Hi adangel
[pmd-unused-private-method.zip](https://github.com/user-attachments/files/18617669/pmd-unused-private-method.zip)

This is a fully compilable project example, only with the classes that I reported.
The classes are together on the same module (and also on same package).
I compiled with:
Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)
Maven home: C:\Java\apache-maven
Java version: 11, vendor: Oracle Corporation, runtime: C:\Java\jdk-11
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

Thanks

**Comment by adangel** [comment]
Hi @manuel-a-romeiro-alb,
thanks for the reproducer example. I can confirm, that this is indeed a problem. The unused private method is wrongly detected by both maven `mvn clean verify` as well as cli `pmd check -d src/main/java --aux-classpath target/classes -R category/java/bestpractices.xml/UnusedPrivateMethod`

This problem doesn't seem to be new: PMD 6.55.0 also detects this.
